### ISSUE TITLE
using hclust inside spark_apply

### ISSUE THREAD

**Issue by sbottelli** [sparklyr/sparklyr#2345]
Hi sparklyr team,
I have an issue using hierchical clustering inside a grouped_by spark_apply.
I would like to using this algorithm for each ID present in my dataset.
If I try to use it with 1 id inside the spark_apply all would be ok, if I try to give as input >=2 different IDs, I receive this error:
20/03/24 16:40:52 **ERROR sparklyr: RScript (7672) terminated unexpectedly: must have n >= 2 objects to cluster** 
20/03/24 16:40:52 ERROR sparklyr: RScript (7672) collected callstack: 
12: stop("must have n >= 2 objects to cluster")
11: hclust(h)
10: (function (tbl) 
{
    library(geosphere)
    library(stats)
    h <- as.dist(distm(tbl[, c("location_longitude", "location_latitude")], 
9: do.call(closure, closure_args)
8: spark_worker_execute_closure(closure, df, funcContext, grouped_by, 
    barrier_map)
7: spark_worker_apply_arrow(sc, config) 

I'm sure that the two IDs have more than two observations, so what could be the problem?

```r
prova <- t_points %>% sparklyr::spark_apply(function(tbl) {
  library(geosphere)
  library(stats)
  h <- as.dist(distm(tbl[,c("location_longitude","location_latitude")], tbl[,c("location_longitude","location_latitude")], fun=distHaversine))
  hc  <- hclust(h)
  clust <- cutree(hc,h=300)
  x <- data.frame("location_datetime" = tbl$location_datetime, "cluster_hspace" = clust)
}, group_by = "date_id_int")
```

**Comment by yl790** [comment]
@sbottelli One limitation of `spark_apply` is the function to be applied cannot reference anything outside of its environment (see https://spark.rstudio.com/guides/distributed-r/#closures) -- although there appears to be no direct reference to external variables in your example, I'm not sure whether there are indirect ones.

Does your example work if you replace spark_apply with lapply?

Also, instead of spark_apply, another alternative to consider (which doesn't have the aforementioned limitation) is using Spark as parallel backend for `foreach` -- this was implemented recently and will be in the next sparklyr release. You can find example usages in https://github.com/sparklyr/sparklyr/blob/master/tests/testthat/test-do-spark.R
The tl;dr is you can simply call `registerDoSpark(sc)` to register doSpark as a parallel backend for `foreach`, and then use `foreach(<my loop variable etc>) %dopar% { <my inner loop> }` to run the inner loop in parallel.

**Comment by sbottelli** [comment]
@yl790 thanks for your great reply!
I solved this issue by using a sdf_repartition of the Spark dataFrame (and using the context to substitute the value inside the function).
Very interesting the new feature to use %foreach%, that could be very very useful.
When the next version will be release? 
Thanks in advance

**Comment by yl790** [comment]
@sbottelli We don't have a definite release date yet. I'll keep you posted.
